### DISCUSSION TITLE
Need to add more features and examples in Table component

### DISCUSSION THREAD

**Discussion by mvmilanjain** [mantinedev/mantine#195] [Feedback]
`Table` is one of the most important component which is required by most of the companies. I have listed some features/examples which you should consider for `Table` component:

- Sorting
- Filtering
- Selection
- Multi Selection
- Inline Editing
- Pagination
- Searching
- Toolbar
- Context Menu
- Async Data Source
- Resizing of Columns
- Fixed Columns
- Scrolling

Please refer [Ant Design Table Component](https://ant.design/components/table/)

**Comment by rtivital** [comment]
Thanks for feedback, Table is a simple wrapper for html `table` element with theme styles, it will remain as is. What you're describing is usually called a DataGrid – I'm planning to add such component in of future releases.

**Reply by mvmilanjain** [reply]
One more suggestion, try to make separate package for datagrid same like dates.

**Reply by rtivital** [reply]
Yep, that was the plan

**Reply by bkniffler** [reply]
What about building the datatable component on top of react-table? Its got so many stuff figured out already, would be mainly about presentational code then @rtivital .

**Reply by rtivital** [reply]
We will evaluate this option before working on data-grid package

**Reply by mvmilanjain** [reply]
> What about building the datatable component on top of react-table? Its got so many stuff figured out already, would be mainly about presentational code then @rtivital .

I have tried Data grid component using React-table & Mantine. Check below ink:
https://codesandbox.io/s/react-table-datagrid-eojw8

**Reply by codellyson** [reply]
this is cool

**Comment by AndersSpringborg** [comment]
Is this a feature you guys want?
I'm writing a version of @mvmilanjain Data grid component, into a core component right now

**Reply by rtivital** [reply]
Yes, this feature is planned.

**Reply by AndersSpringborg** [reply]
Okay. Are you okay with added react table as a dependency, to mantine/core?

**Reply by rtivital** [reply]
DataGrid should be a separate package `@mantine/data-grid`

**Reply by mvmilanjain** [reply]
I have tried my hand with React-Table library.
Check out link: https://codesandbox.io/s/react-table-datagrid-eojw8

**Comment by patrikelfstrom** [comment]
Here is a good read about the differences between a table and a data grid:
https://blog.ag-grid.com/react-data-grid-vs-react-data-table-vs-react-grid/

One important thing I think is that it's recommended to build a data grid using divs instead of tables.
For example React Virtualized/window, AG Grid, React Data Grid and MUI Data Grid all use divs instead of tables.

**Reply by emilhe** [reply]
I am also really interested in a mantine `DataGrid` component. I was considering implementing one myself, but if it is already on the roadmap, I guess it would be better to wait for an official component :)

**Comment by bogordesaincom** [comment]
i have using with react table 
<img width="1438" alt="Jepretan Layar 2022-03-17 pukul 15 07 31" src="https://user-images.githubusercontent.com/51335197/158764769-f3f78523-b782-454a-8c32-f18b8f7415c1.png">

**Reply by hunt3r** [reply]
Did you just add manual styles or did you find some table styles play well with react table. I'm also using this approach but react table doesn't give really any styling guidelines.

**Reply by zeestack** [reply]
react-table is headless, it will not provide you any styling. You can style your table the way you want.

**Comment by nealoke** [comment]
Any update on this? :) Seems like a very nice addition

**Reply by rtivital** [reply]
We are not planning to add this component any time soon.

**Reply by elarens** [reply]
Sadly, it makes Mantine unusable for me (and likely many other frontend developers specialising in admin interfaces). Tables (or DataGrids, if you prefer) are the core components for managing complex data, and not having basic capabilities like sorting, filtering, selection, search, pagination and async data fetching means that I have to choose another component library, as it is too time-consuming to build those things from scratch. That's unfortunate, as I was getting very impressed with Mantine otherwise.

**Reply by bkniffler** [reply]
As much as I miss a table component, I find its not too hard to implement your own (as shown by @mvmilanjain above, using react-table). Maybe someone in the community will create a table component and publish as a separate lib. And maybe this could get merged into mantine at some point.

**Reply by clayreinhardt** [reply]
ugh

**Comment by hunt3r** [comment]
I’ve been using react table for this. It’s super bare bones on styling and is pretty agnostic on markup so it works fairly well with mantine. Albeit a bit more work to implement. 

I agree that a more robust out of the box solution would be ideal. 

On May 11, 2022, at 8:00 AM, Benjamin Kniffler ***@***.***> wrote:

﻿
As much as I miss a table component, I find its not too hard to implement your own (as shown by @mvmilanjain above, using react-table). Maybe someone in the community will create a table component and publish as a separate lib. Any maybe this could get merged into mantine at some point.

**Comment by icflorescu** [comment]
You might want to take a look at [Mantine DataTable](https://icflorescu.github.io/mantine-datatable/), a component I've built with Mantine primitives (not React Table), that covers most of the features mentioned above and should support:

- Asynchronous data loading
- Pagination
- Multiple rows selection
- Scrollable behavior, while maintaining a fixed header, pagination footer and row selection column
- Column sorting
- Custom cell data rendering
- Row context-menu
- Dark themes

I've also tried to build a nice documentation website (following as much as possible the excellent Mantine's docs design patterns) and I've included a series of examples and usage scenarios.
The component properties are be self explanatory, well typed and also documented with JSDoc, so it should be quite straightforward to use.

[![Mantine DataTable](https://user-images.githubusercontent.com/581999/187723353-0e082dbc-925b-4d0d-a14e-94c00073ee08.png)](https://icflorescu.github.io/mantine-datatable/)

**Reply by mvmilanjain** [reply]
Very nice implementation

**Reply by wachidmudi** [reply]
Is it will have fixed column feature like [rc-table](https://table-react-component.vercel.app/demo/fixed-columns)?

**Reply by icflorescu** [reply]
@wachidmudi
So far only the row selection column is fixed, but maybe we could do something about this in a future version.
PRs are welcomed! ;-)

**Reply by wachidmudi** [reply]
Okay. Nice work anyway.

**Reply by elarens** [reply]
@icflorescu Beautiful implementation. Would be great to have a render prop to render ReactNode in cells instead of just text, much like Ant Design tables do, since it's essential for functionality like rendering action buttons that then allow users to further interact with the data. But very nice work otherwise.

**Reply by icflorescu** [reply]
@elarens Thank you for the kind words! A [Mantine DataTable](https://icflorescu.github.io/mantine-datatable/) column **DOES** have a `render` prop that allows you to output ReactNode in cells:
```ts
/**
 * Custom cell data render function accepting the current record
 */
render?: (record: T) => ReactNode;
```
Have a look at the example in the [getting started](https://icflorescu.github.io/mantine-datatable/getting-started) page.
Perhaps I should make this more evident in the [column properties example page](https://icflorescu.github.io/mantine-datatable/examples/column-properties), or even create a specifically-targeted example...

**Reply by elarens** [reply]
@icflorescu Indeed it is there. I missed it, apologies - should have been more attentive. You've done an incredible job.

**Reply by icflorescu** [reply]
Starting with `v.1.7.0`, Mantine DataTable features [row expansion](https://icflorescu.github.io/mantine-datatable/examples/expanding-rows).

**Comment by hunt3r** [comment]
Awesome, I think this will work out nicely to replace the react table stuff
I have. I'm going to give it a try.

**Comment by petrvecera** [comment]
Hello guys

We are evaluating Mantine for our next big project. It looks like awesome library which we would like to use. And  it looks like this is basically the only thing which is stopping us. The https://github.com/icflorescu/mantine-datatable looks really great. But we would like to avoid adding 3rd party components. You usually want to keep your dependencies to minimum or you will have problems 3 years later...  

Is there any update on the Mantine table inside the lib? Maybe even in beta-version? Because it might takes few months before we hit production. So even beta version would be fine for us right now. 

Thank you 
🚀 ❤️

**Reply by rtivital** [reply]
There are no plans of adding data grid component soon. You can use one of the libraries built by community.

**Comment by KevinVandy** [comment]
Here's another 3rd party Mantine Table library I'm introducing as an option here: www.mantine-react-table.com/

**Comment by RabeeAbuBaker** [comment]
Hi @rtivital, any updates on the data grid component?

**Reply by rtivital** [reply]
I'm not planning to work on it any time soon, you can use one of the packages provided by the community.

**Reply by ctretyak** [reply]
and we used mantine-react-table and now we cannot update to mantine@8 :(

**Comment by langziyang** [comment]
hope make table or datagrid is responsive

**Comment by kevin-griley** [comment]
Just noting I agree to use TanStack Table for this in the future, its headless and already works will with mantine.

**Comment by icflorescu** [comment]
Just wanted to let everyone interested know:

### [Mantine DataTable V7](https://icflorescu.github.io/mantine-datatable/) is here and fully supports Mantine V7 🎉

![Mantine DataTable V7](https://github.com/icflorescu/mantine-datatable/assets/581999/0be62c66-978e-4ecc-9abe-0dad3254c94d)

Be advised that it comes with a few [breaking changes](https://github.com/icflorescu/mantine-datatable/blob/main/CHANGELOG.md#710-alpha1-2023-10-30), as expected.
The amount of effort spent to rewrite the entire documentation website with Next.js 14 was gruesome, but I had to do it in order to make sure the component works in every example.

If everything go as planned and we're not bumping into any critical bugs, I'm planning to release the final version in a day or two.
